url0d38a10ee7_diegopliebana_GEBT_tgz-pJ8-ch_idsia_scenarios_test_EvolveWithChangingSeedsJ8/src/ch/idsia/mario/engine/sprites/Mario.java

```diff
@@ -0,0 +1,815 @@
+
+        byte block = world.level.getBlock(x, y);
+
+        if (((Level.TILE_BEHAVIORS[block & 0xff]) & Level.BIT_PICKUPABLE) > 0)
+        {
+            Mario.getCoin();
+            world.level.setBlock(x, y, (byte) 0);
+            for (int xx = 0; xx < 2; xx++)
+                for (int yy = 0; yy < 2; yy++)
+                    world.addSprite(new Sparkle(x * 16 + xx * 8 + (int) (Math.random() * 8), y * 16 + yy * 8 + (int) (Math.random() * 8), 0, 0, 0, 2, 5));
+        }
+
+        if (blocking && ya < 0)
+        {
+            world.bump(x, y, large);
+        }
+
+        return blocking;
+    }
+
+    public void stomp(Enemy enemy)
+    {
+        if (deathTime > 0 || world.paused) return;
+
+        float targetY = enemy.y - enemy.height / 2;
+        move(0, targetY - y);
+
+        xJumpSpeed = 0;
+        yJumpSpeed = -1.9f;
+        jumpTime = 8;
+        ya = jumpTime * yJumpSpeed;
+        onGround = false;
+        sliding = false;
+        invulnerableTime = 1;
+    }
+
+    public void stomp(Shell shell)
+    {
+        if (deathTime > 0 || world.paused) return;
+
+        if (keys[KEY_SPEED] && shell.facing == 0)
+        {
+            carried = shell;
+            shell.carried = true;
+            setRacoon(true);
+        }
+        else
+        {
+            float targetY = shell.y - shell.height / 2;
+            move(0, targetY - y);
+
+            xJumpSpeed = 0;
+            yJumpSpeed = -1.9f;
+            jumpTime = 8;
+            ya = jumpTime * yJumpSpeed;
+            onGround = false;
+            sliding = false;
+            invulnerableTime = 1;
+        }
+    }
+
+    public void getHurt()
+    {
+        if (deathTime > 0 || world.paused || isMarioInvulnerable) return;
+
+        if (invulnerableTime > 0) return;
+
+        if (large)
+        {
+            world.paused = true;
+            powerUpTime = -3 * FractionalPowerUpTime;
+            if (fire)
+            {
+                world.mario.setLarge(true, false);
+            }
+            else
+            {
+                world.mario.setLarge(false, false);
+            }
+            invulnerableTime = 32;
+        }
+        else
+        {
+            die();
+        }
+    }
+
+    public void win()
+    {
+        xDeathPos = (int) x;
+        yDeathPos = (int) y;
+        world.paused = true;
+        winTime = 1;
+        status = Mario.STATUS_WIN;
+    }
+
+    public void die()
+    {
+        xDeathPos = (int) x;
+        yDeathPos = (int) y;
+        world.paused = true;
+        deathTime = 25;
+        status = Mario.STATUS_DEAD;
+    }
+
+
+    public void getFlower()
+    {
+        if (deathTime > 0 || world.paused) return;
+
+        if (!fire)
+        {
+            world.paused = true;
+            powerUpTime = 3 * FractionalPowerUpTime;
+            world.mario.setLarge(true, true);
+        }
+        else
+        {
+            Mario.getCoin();
+        }
+        ++flowersDevoured;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 704:716, 721:733.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.